### PR TITLE
add Clipboard component for ios and android

### PR DESCRIPTION
**Pull request by TanTan-TT**
add Clipboard component for ios and android
 ```javascript
    import Clipboard from 'react-native'

    Clipboard.get((content)=>{
          console.log('here is content in clipboard:%s',content)
    });
    var content = 'here is a string';
    Clipboard.set(content);
```

### CODE REVIEW

**Comment by satya164** [comment]
Hey. Thanks for working on this.

There is already a `Pasteboard` component for iOS - https://github.com/facebook/react-native/blob/master/React/Modules/RCTPasteboard.m

For Android, you might want to follow the same API.

**Comment by ide** [comment]
cc @dmmiller in case an Android version of Pasteboard is in the works.

**Comment by TanTan-TT** [comment]
@satya164 I know it, but you don't provide js interface. And the version of iOS has not "get" method

**Comment by satya164** [comment]
@TanTan-TT Yes, but it's called `Pasteboard` and not `Clipboard`.  And the method is probably called `setString`. We shouldn't change the whole API.

**Comment by TanTan-TT** [comment]
@satya164 I change the API
```javascript
import Pasteboard from 'react-native'

Pasteboard.getString((content)=>{
      console.log('here is content in pasteboard:%s',content)
});
var content = 'here is a string';
Pasteboard.setString(content);
```

**Comment by satya164** [comment]
Let's wait for @dmmiller

**Comment by mkonicek** [comment]
There's already the iOS Pasteboard module. Why are you adding Libraries/Components/Pasteboard/Pasteboard.ios.js?

File: ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * {@link NativeModule} that allows JS to get content in clipboard.
+ */
+public class PasteboardModule extends ReactContextBaseJavaModule {
+    public PasteboardModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+    }
+
+    @Override
+    public String getName() {
+        return "PasteboardAndroid";
```
Comment (dmmiller) [inline ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java]: If you made this just RCTPasteboard, you could get away with one Pasteboard,js file that referenced the native modules.

**Comment by nicklockwood** [comment]
Agree with @dmmiller. We're trying to unify iOS and Android APIs as much as possible, and I don't see a good reason to name these differently.

The iOS Pasteboard module is pretty new, and doesn't have a JS wrapper yet, so I'm cool with calling the JS file "Clipboard.js" if people think that's a better name - I can rename the iOS native module later.

As for the missing get method on iOS, just mark that API as Android-only for now and we can add iOS support later.

**Comment by ide** [comment]
Clipboard is a more common term I think: https://www.google.com/trends/explore#q=clipboard%2C%20pasteboard&cmpt=q&tz=Etc%2FGMT%2B8

**Comment by satya164** [comment]
Yeah, `Clipboard` is a better name in my opinion too.

**Comment by TanTan-TT** [comment]
@dmmiller sorry, I don't understand your solution, do you mean if There isn't a js wrapper for pasteboard, directly use methods exported from RTCPasteboard? 
But in my opinion, if do this, users must import NativeModule instead of react-native?
And I think in Java world RTCPasteboard is a strange class name

**Comment by TanTan-TT** [comment]
@ide @satya164 maybe I could change name of js wrapper to "Clipboard ", then users can import it from "react-native"

**Comment by nicklockwood** [comment]
@TanTan-TT that sounds like a good approach. Name the JS file and the android class "Clipboard", leave the iOS class as Pasteboard in iOS and in Clipboard.js file do something like:

    var RCTClipboard = NativeModules.Clipboard || NativeModules.Pasteboard;

There's no need to add a Pasteboard.ios.js or anything like that.

File: ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            ClipboardManager clipboard = (ClipboardManager)reactContext.getSystemService(reactContext.CLIPBOARD_SERVICE);
+
+            ClipData clip = ClipData.newPlainText("text", data);
+
+            clipboard.setPrimaryClip(clip);
+
+        } catch(Exception e) {
+            Log.w("PasteboardModule", "Error in SET operation: " + e.getMessage());
+        }
+    }
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java]: Please remove all the empty lines :)

File: ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            ReactApplicationContext reactContext = getReactApplicationContext();
+
+            ClipboardManager clipboard = (ClipboardManager)reactContext.getSystemService(reactContext.CLIPBOARD_SERVICE);
+
+            ClipData clip = ClipData.newPlainText("text", data);
+
+            clipboard.setPrimaryClip(clip);
+
+        } catch(Exception e) {
+            Log.w("PasteboardModule", "Error in SET operation: " + e.getMessage());
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java]: We have to use `Flog` in React Native.

**Comment by mkonicek** [comment]
We do have a pasteboard module internally and plan to open source it:
https://gist.github.com/mkonicek/4556908b269919e1bd31

Please base your code on that.

I agree with others that Clipboard is a much better name both for Android and iOS. Please replace all the occurrences to "pasteboard" you can find, including string constants, Java class and package names, etc.

File: ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java (hunk 2 above)
Comment (TanTan-TT) [inline ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java]: I am out of the office until 12/06/2015.

For any issues, plz ask Jinhao.Geng for help.


Note: This is an automated response to your message  "Re: [react-native]
add Clipboard component for ios and android (#4384)" sent on 12/3/2015
7:04:32 AM.

This is the only notification you will receive while this person is away.

File: ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java (hunk 3 above)
Comment (TanTan-TT) [inline ReactAndroid/src/main/java/com/facebook/react/modules/pasteboard/PasteboardAndroid.java]: I am out of the office until 12/06/2015.

For any issues, plz ask Jinhao.Geng for help.


Note: This is an automated response to your message  "Re: [react-native]
add Clipboard component for ios and android (#4384)" sent on 12/3/2015
7:05:03 AM.

This is the only notification you will receive while this person is away.

**Comment by TanTan-TT** [comment]
I am out of the office until 12/06/2015.

For any issues, plz ask Jinhao.Geng for help.


Note: This is an automated response to your message  "Re: [react-native]
add Clipboard component for ios and android (#4384)" sent on 12/3/2015
7:10:21 AM.

This is the only notification you will receive while this person is away.

**Comment by TanTan-TT** [comment]
@mkonicek OK, I will deal with it. But I still have a suggestion about js wrapper file.
Based on v0.16's pasteboard,  users must use this module like this 
```javascript
var pasteboard  = require('NativeModules').pasteboard;
```
But for other API, code is like
```javascript
import {ToastAndroid} from 'react-native'
```
I think above code is more formal than require('NativeModule'), which is like an internal API.

**Comment by satya164** [comment]
Yeah, I agree that requiring the mdule should be consistent with other things like ToastAndroid.

**Comment by TanTan-TT** [comment]
@mkonicek I have pulled request the newest version(include adding a demo in UIExplore) and I have passed the test both in iOS and Android UIExplore.
But Travis CI told me there are some errors,
```
Examples/UIExplorer/ClipboardExample.js:20
 20:    Clipboard,
        ^^^^^^^^^ property `Clipboard`. Property not found in
 23:  } = React;
          ^^^^^ Object.create
````
But indeed it works. I don't know how to handle these errors, and I believe some errors doesn't caused by my code.  But still you can review codes.
Thank you.

**Comment by satya164** [comment]
Sorry for bringing up yet another thing in the already long opened PR, but do we really want the method names to be `Clipboard.getClipboardString` and `Clipboard.setClipboardString`? Personally, I prefer `Clipboard.getString` and `Clipboard.setString`, no need for the extra `Clipboard`.

**Comment by TanTan-TT** [comment]
@satya164 In js API, it indeed is no "Clipboard"

**Comment by satya164** [comment]
@TanTan-TT Is there any reason it's different in Native?

File: Libraries/Components/Clipboard/Clipboard.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule Clipboard
+ */
+'use strict';
+
+var NativeModules = require('NativeModules');
+var NativeClipboard = NativeModules.Clipboard;
+if(!NativeClipboard){
```
Comment (satya164) [inline Libraries/Components/Clipboard/Clipboard.js]: Why is this check necessary? Where is `ClipboardAndroid`?

**Comment by satya164** [comment]
If we change the Native methods to `setString` and `getString`, we can greatly simplify the `Clipboard.js` file,

```js
module.exports = require('NativeModules').Clipboard;
```

**Comment by TanTan-TT** [comment]
@satya164 both ios and Android?

**Comment by satya164** [comment]
@TanTan-TT Yeah :)

**Comment by TanTan-TT** [comment]
@satya164 Fix it!

**Comment by nicklockwood** [comment]
@facebook-github-bot shipit

File: Libraries/BatchedBridge/BatchedBridgedModules/__mocks__/NativeModules.js [hunk 5]
```diff
@@ -54,8 +54,8 @@ var NativeModules = {
   AlertManager: {
     alertWithArgs: jest.genMockFunction(),
   },
-  Pasteboard: {
-    setPasteboardString: jest.genMockFunction(),
+  Clipboard: {
+    setClipboardString: jest.genMockFunction(),
```
Comment (satya164) [inline Libraries/BatchedBridge/BatchedBridgedModules/__mocks__/NativeModules.js]: Seems you forgot this one :)

Comment (nicklockwood) [inline Libraries/BatchedBridge/BatchedBridgedModules/__mocks__/NativeModules.js]: It's alright, I've fixed in at our end.

Comment (satya164) [inline Libraries/BatchedBridge/BatchedBridgedModules/__mocks__/NativeModules.js]: @nicklockwood Cool :D

**Comment by TanTan-TT** [comment]
@nicklockwood thx

**Comment by mkonicek** [comment]
Merging this. I ended up tweaking the Java code a bit.

**Comment by mkonicek** [comment]
Looking at the API again, I really think `getString` should provide an error callback, rather than failing silently if it didn't work. Can you send a separate PR for that once this lands?

**Comment by TanTan-TT** [comment]
@mkonicek OK. Got it.

**Comment by sunnylqm** [comment]
Can someone make a doc for this please? Thanks!

**Comment by TanTan-TT** [comment]
@sunnylqm I can write, but I don't know how to write this in this WebSite. @satya164 Can you guide me to write a document for Clipboard?

**Comment by sunnylqm** [comment]
They seems to put doc in a certain format in the component source, like javadoc.

**Comment by TanTan-TT** [comment]
@sunnylqm I know it. But in this scenario, there is not js file and no place for jsdoc.

**Comment by satya164** [comment]
@TanTan-TT You can write the APIs inside the JS file. We'll need to do that for backward compatibility anyways.

**Pull request closed without merging**